### PR TITLE
chore(dataset/array): Add Bitpacked encoding

### PR DESCRIPTION
**Pull request by rfratto**
Adds a Bitpacked encoding to dataset/array, which encodes a batch of numbers into the smallest possible set of bits.

---

> [!NOTE]
> **Medium Risk**
> Introduces a new block-based integer encoding/decoding path with custom bit packing/unpacking logic and new child-array metadata, which could impact correctness/performance for encoded datasets despite good test coverage.
> 
> **Overview**
> Adds a new `bitpacked` array encoding for `uint32`/`uint64` that stores values in fixed-size blocks using the minimum bit-width per block, plus a widths child array (and optional validity child array for nullable types).
> 
> `NewWriter`/`NewReader` now route `EncodingKindBitpacked` to new bitpacking codec implementations, and the `memory.Buffer` API gains `Next(n)` to support efficient incremental writes. Includes extensive unit tests and a benchmark covering nullable/non-nullable cases and edge conditions (all zeros, max width, mixed-width blocks).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f5ae3465cd7f37dd6807c586d70fbd2cdd0b0049. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: pkg/dataset/array/codec_bitpacked.go [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+		return io.EOF
+	}
+
+	enc := r.arr.Encoding.(*EncodingBitpacked)
+	width := int(r.widths[r.blockIndex])
+	blockRows := min(enc.BlockSize, r.arr.RowCount-r.rowOff)
+
+	r.blockBuf.Grow(blockRows)
+	r.blockBuf.Resize(blockRows)
+	r.blockBuf.Clear()
```
[the remark this answers is not in the document]
Comment (rfratto) [reply pkg/dataset/array/codec_bitpacked.go:386]: This is assuming bytes.Buffer API, which is not what this is. Invalid comment

**Review by benclive — DISMISSED** [review]

**Review by benclive — APPROVED** [review]
